src/cpu/flexcpu/simple_dataflow_thread.hh

```diff
@@ -0,0 +1,154 @@
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: Bradley Wang
+ */
+
+#ifndef __SIMPLE_DATAFLOW_CPU_THREAD_HH__
+#define __SIMPLE_DATAFLOW_CPU_THREAD_HH__
+
+#include <vector>
+
+#include "cpu/exec_context.hh"
+#include "cpu/flexcpu/simple_dataflow_cpu.hh"
+#include "cpu/simple_thread.hh"
+
+
+/**
+ * Instances of this class represent a logical "Core" within the
+ * SimpleDataflowCPU.
+ */
+class SDCPUThread
+{
+  protected:
+    // BEGIN Internal class definitions
+
+    class SDCPUExecContext : public ExecContext
+    {
+      protected:
+        SDCPUThread& backingState;
+        
+      public:
+        SDCPUExecContext(SDCPUThread& backing_state):
+            backingState(backing_state)
+        { }
+    };
+
+    class DynamicInst
+    {
+        StaticInst& instructionRef;
+        // TODO instruction tag
+        // TODO instruction results
+    };
+
+    // END Internal class definitions
+
+    SimpleDataflowCPU* _cpuRef;
+
+    /**
+     * Monotonically increasing count of the instructions executed within this
+     * thread thus far. Since it should be unique for every (committed)
+     * execution of an instruction, this should be freely usable as a tag for
+     * our dynamic instructions.
+     */
+    typedef uint64_t InstTag;
+    InstTag numInstsCommitted = 0;
+    InstTag numInstsSpeculated = 0;
+
+    /**
+     * The SimpleThread class
+     */
+    SimpleThread _committedState;
+
+    // BEGIN Speculative state
+
+    /**
+     * Instruction execution will pass through an ExecContext interface,
+     * therefore in most cases, it will be tied to speculative state, instead
+     * of the "real" state of the system.
+     */
+    SDCPUExecContext _isaInterface;
+
+    TheISA::PCState _speculativePC;
```

Comment:
We might want to have more than one speculative PC. Maybe not, though. Something to keep in mind.